pkg/admission/prep_downscale.go

```diff
@@ -0,0 +1,208 @@
+			Result: &metav1.Status{
+				Message: fmt.Sprintf("downscale of %s/%s in %s from %d to %d replicas is not allowed because the %v label is not set or empty.", ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas, PrepDownscalePortKey),
+			},
+		}
+		level.Warn(logger).Log("msg", fmt.Sprintf("downscale not allowed because the %v label is not set or empty", PrepDownscalePortKey))
+		return &reviewResponse
+	}
+
+	path, ok := lbls[PrepDownscalePathKey]
+	if !ok || len(path) == 0 {
+		reviewResponse := v1.AdmissionResponse{
+			Allowed: false,
+			Result: &metav1.Status{
+				Message: fmt.Sprintf("downscale of %s/%s in %s from %d to %d replicas is not allowed because the %v label is not set or empty.", ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas, PrepDownscalePathKey),
+			},
+		}
+		level.Warn(logger).Log("msg", fmt.Sprintf("downscale not allowed because the %v label is not set or empty", PrepDownscalePathKey))
+		return &reviewResponse
+	}
+
+	// Since it's a downscale, check if the resource has the label that indicates it needs to be prepared to be downscaled.
+	// Create a slice of endpoint addresses for pods to send HTTP post requests to and to fail if any don't return 200
+	if !*ar.Request.DryRun {
+		diff := (*oldReplicas - *newReplicas)
+		type endpoint struct {
+			url   string
+			index int
+		}
+		eps := make([]endpoint, diff)
+
+		// The DNS entry for a pod of a stateful set is
+		// ingester-zone-a-0.$(servicename).$(namespace).svc.cluster.local
+		// The service in this case is ingester-zone-a as well.
+		// https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id
+		for i := 0; i < int(diff); i++ {
+			index := int(*oldReplicas) - i - 1 // nr in statefulset
+			eps[i].url = fmt.Sprintf("%v-%v.%v.%v.svc.cluster.local:%s/ingester/%s",
+				ar.Request.Name, // pod name
+				index,
+				ar.Request.Name, // svc name
+				ar.Request.Namespace,
+				port,
+				path,
+			)
+			eps[i].index = index
+		}
+
+		g, _ := errgroup.WithContext(ctx)
+		for _, ep := range eps {
+			ep := ep // https://golang.org/doc/faq#closures_and_goroutines
+			g.Go(func() error {
+				logger = log.With(logger, "url", ep.url, "index", ep.index)
+				resp, err := client.Post("http://"+ep.url, "application/json", nil)
+				if err != nil {
+					level.Error(logger).Log("msg", "error sending HTTP post request", "err", err)
+					return err
+				}
+				if resp.StatusCode != 200 {
+					err := errors.New("HTTP post request returned non-200 status code")
+					body, readError := io.ReadAll(resp.Body)
+					defer resp.Body.Close()
+					level.Error(logger).Log("msg", err, "status", resp.StatusCode, "response_body", body)
+					return errors.Join(err, readError)
+				}
+				level.Debug(logger).Log("msg", "pod prepared for shutdown")
+
+				err = addPreparedForDownscaleAnnotationToPod(ctx, api, ar.Request.Namespace, ar.Request.Name, ep.index)
```

Comment:
I think the annotation should be added to the pod _before_ the endpoint is called. This avoids a potential race condition:
* The "prep shutdown" endpoint is called, the pod / ingester in question modifies its in-memory configuration accordingly.
* The periodic task that uses the [downwardAPI](https://kubernetes.io/docs/concepts/workloads/pods/downward-api/) to read annotations from a file on disk runs after the "prep shutdown" endpoint is called but before an annotation has been added to the pod, reverting the configuration change made by the "prep shutdown" endpoint.

Changing the order to add the annotation first and then call the endpoint means that the result of calling the endpoint won't be reverted by the periodic job.

Reply:
Good point. This was added.